experimental/uefi/channel/src/lib.rs

```diff
@@ -32,25 +33,76 @@ pub mod schema {
-pub type FrameLength = u32;
+type FrameLength = u32;
 const FRAME_LENGTH_SIZE: usize = 4;
 static_assertions::assert_eq_size!([u8; FRAME_LENGTH_SIZE], FrameLength);
 
-/// In request frames this field contains the id of the method_or_status that is invoked.
-/// In response frames this field contains the status code of the method_or_status invocation.
-pub type Method = u32;
+/// In request frames this field contains the id of the method that is invoked.
+/// In response frames this field contains the status code of the method invocation.
+type MethodOrStatus = u32;
 const METHOD_SIZE: usize = 4;
-static_assertions::assert_eq_size!([u8; METHOD_SIZE], Method);
+static_assertions::assert_eq_size!([u8; METHOD_SIZE], MethodOrStatus);
 
-pub const FRAME_HEADER_SIZE: usize = 8;
+// TODO(#2848): Use this space to send a stream identifier that identifies
+// frames as belonging to the same message. Currently the frames of every message
+// must be sent as an uninterrupted seqeuence. A stream identifier would allow us
+// to interleave messages.
+/// Reserved space to maintain 8 byte alignement.
+const SPACER_SIZE: usize = 4;
+
+#[repr(u32)]
+#[derive(Copy, Clone)]
+enum Flag {
+    /// Atomic frame, a message sent as a single frame
+    AtomicFrame = 0,
+    /// Starts a stream of frames that constitute a message
+    StreamStart = 1,
+    /// Continues a stream of frames that constitute a message
+    StreamContinuance = 2,
+    /// Ends a stream of frames that constitute a message, indicating that the recipient may
+    /// process the message.
+    StreamEnd = 3,
+}
+const FLAG_SIZE: usize = 4;
+static_assertions::assert_eq_size!([u8; FLAG_SIZE], Flag);
+impl TryFrom<[u8; FLAG_SIZE]> for Flag {
+    type Error = anyhow::Error;
+
+    fn try_from(value: [u8; FLAG_SIZE]) -> Result<Self, Self::Error> {
+        let flag = match u32::from_le_bytes(value) {
+            0 => Flag::AtomicFrame,
+            1 => Flag::StreamStart,
+            2 => Flag::StreamContinuance,
+            3 => Flag::StreamEnd,
+            _ => anyhow::bail!("invalid flag"),
+        };
+        Ok(flag)
+    }
+}
+impl From<Flag> for [u8; 4] {
+    fn from(flag: Flag) -> Self {
+        (flag as u32).to_le_bytes()
+    }
+}
+
+const FRAME_HEADER_SIZE: usize = 16;
 static_assertions::assert_eq_size!(
     [u8; FRAME_HEADER_SIZE],
-    [u8; FRAME_LENGTH_SIZE + METHOD_SIZE]
+    [u8; FRAME_LENGTH_SIZE + METHOD_SIZE + SPACER_SIZE + FLAG_SIZE]
 );
+// Ensure that the frame header is 8 Byte aligned.
+static_assertions::const_assert!(FRAME_HEADER_SIZE % 8 == 0);
 
-pub struct Frame {
-    pub method_or_status: Method,
-    pub body: Vec<u8>,
+const MAX_FRAME_SIZE: usize = 64000;
+const MAX_FRAME_BODY_SIZE: usize = MAX_FRAME_SIZE - FRAME_HEADER_SIZE;
+
+/// A [`Frame`] is a small unit data that is sent over the communication
+/// channel. Usually several [`Frame`]'s are used to send a [`Message`].
+#[derive(Clone)]
+struct Frame {
+    method_or_status: MethodOrStatus,
```

Comment:
In fact I think in principle the framing layer should not even know what the message represents (and in particular whether it has a method or status or anything else). It should just get some bytes, and split them into fragments, and then reconstruct them on the other side.

Reply:
I don't disagree, logical layers seem nice. But IRRC we had this discussion at once point earlier, and decided that for now it would be nice to have the method recorded in each frame, in case it helps with routing / prioritization later on. 

Happy to revisit this, but for now it might be good to keep it around. :)

Reply:
Another reason to favour removing the method ID from the header is that (as far as I can tell) is the only reason for having the `PartialMessage` construct. But I agree that it is fine to keep as is for now and refactor as needed as part of future code changes (will probably happen naturally as part of supportring multiplexed/interleaves message).

Reply:
Yes, since we're dealing with one message at a time we could move all logic into a single function that uses local state only. 

It seems that's not necessarily tied to the `method_id`, said function with local state could likewise check that this value is consistent.

Reply:
I should note that if we do go with buffer approach we will require two IDs, one to link request messages with their response messages, and another one to link frames belonging to the same message. 

Imo fine and nice for having logical layers, but I recall that we earlier felt this was undesireable.

Reply:
To be fair, that comment was under the assumption that the initial implementation would not use the IDL, and that having a single header would simplify the code and the initial implementation. At this point it seems that having a single header is complicating the code rather than simplifying it.

For this PR, I am fine with either approach. We will probably go through at least one more design cycle around the IDL and framing to implement interleaving, so any refactoring can happen as part of that.

Reply:
Makes sense, thanks for explaining :)

Reply:
Updated the design doc with the new approach of using clearly delineated layers (and by extension seperate IDs for frames and messages). https://docs.google.com/document/d/1Q4u5hYzOFF74_RfwZzpRWgSSwJwfQ-GXRvNllgCXOO8/edit?usp=sharing